### PR TITLE
Step 8

### PR DESCRIPTION
**Pull request by drozddmitriy**

### CODE REVIEW

File: spec/features/sign_up_page_spec.rb [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+RSpec.describe 'Registration', type: :feature do
+  let!(:user) { create(:user) }
+
+  before { visit new_user_registration_path }
+
+  context 'when account exists' do
+    before do
+      fill_in 'Email', with: user.email
+      fill_in 'Password', with: user.password
```
Comment (za4emnik) [inline spec/features/sign_up_page_spec.rb]: i18n

File: app/models/line_item.rb [hunk 2]
```diff
@@ -0,0 +1,10 @@
+class LineItem < ApplicationRecord
```
Comment (za4emnik) [inline app/models/line_item.rb]: Why LineItem? Maybe OrderItem?

File: app/admin/books.rb [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+    end
+    column :category
+    column :authors do |resource|
+      BookDecorator.new(resource).author_full_name
+    end
+    column :description do |resource|
+      truncate(resource.description, length: 250)
+    end
+    column :price do |resource|
+      number_to_currency(resource.price, unit: I18n.t('currency'))
```
Comment (epoberezhny) [inline app/admin/books.rb]: Config I18n https://github.com/svenfuchs/rails-i18n/blob/master/rails/locale/en.yml

File: app/admin/books.rb [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+
+    active_admin_comments
+  end
+
+  form do |f|
+    f.semantic_errors(*f.object.errors.keys)
+
+    f.inputs do
+      f.input :title
+      f.input :authors, as: :check_boxes, collection: AuthorDecorator.decorate_collection(Author.all)
```
Comment (epoberezhny) [inline app/admin/books.rb]: Please take a look https://activeadmin.info/11-decorators.html

File: app/admin/books.rb [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+      f.input :dimension_w
+      f.input :dimension_d
+      f.input :quantity
+    end
+
+    f.inputs do
+      f.input :images, as: :file, input_html: { multiple: true }
+    end
+
+    f.inputs 'Images', multipart: true do
```
Comment (epoberezhny) [inline app/admin/books.rb]: I18n

File: app/admin/coupons.rb [hunk 6]
```diff
@@ -0,0 +1,14 @@
+ActiveAdmin.register Coupon do
+  preserve_default_filters!
+  remove_filter :order
+  permit_params :coupon, :active, :sale
+
+  form do |f|
+    f.inputs do
+      f.input :coupon, input_html: { value: CouponService.generate }
+      f.input :active, input_html: { checked: true }
+      f.input :sale, label: 'Sale'
```
Comment (epoberezhny) [inline app/admin/coupons.rb]: I18n

File: app/assets/javascripts/checkbox_input.js [hunk 7]
```diff
@@ -0,0 +1,4 @@
+function checkForm(form) {
+ if (form.check.checked) form.commit.disabled = 0
```
Comment (epoberezhny) [inline app/assets/javascripts/checkbox_input.js]: Make it beautiful

File: app/controllers/order_items_controller.rb [hunk 8]
```diff
@@ -0,0 +1,42 @@
+class OrderItemsController < ApplicationController
+  load_and_authorize_resource
+
+  def index
+    @items = current_order.order_items.includes(:book)
+  end
+
+  def create
+    @order_item = OrderItem.find_or_initialize_by(book_id: params[:order_item][:book_id], order_id: current_order.id)
```
Comment (epoberezhny) [inline app/controllers/order_items_controller.rb]: Why do we retrieve parameters from params explicitly?

File: app/controllers/pages_controller.rb [hunk 9]
```diff
@@ -0,0 +1,8 @@
+class PagesController < ApplicationController
+  authorize_resource class: false
```
Comment (epoberezhny) [inline app/controllers/pages_controller.rb]: Wgy do we need this?

File: app/controllers/order_items_controller.rb [hunk 10]
```diff
@@ -0,0 +1,41 @@
+class OrderItemsController < ApplicationController
+  load_and_authorize_resource
```
Comment (epoberezhny) [inline app/controllers/order_items_controller.rb]: except: [:index]

File: app/controllers/users/sessions_controller.rb [hunk 11]
```diff
@@ -0,0 +1,24 @@
+class Users::SessionsController < Devise::SessionsController
+  def new
+    super
+  end
+
+  def create
+    super
+  end
```
Comment (epoberezhny) [inline app/controllers/users/sessions_controller.rb]: Why???

File: app/decorators/book_decorator.rb [hunk 12]
```diff
@@ -0,0 +1,15 @@
+class BookDecorator < Draper::Decorator
+  delegate_all
+
+  def author_full_name
+    authors.map { |author| author.first_name + ' ' + author.last_name }.join(', ')
+  end
+
+  def book_dimensions
+    "H: #{dimension_h} x W: #{dimension_w} x D: #{dimension_d}"
```
Comment (epoberezhny) [inline app/decorators/book_decorator.rb]: I18n

File: app/decorators/order_addresses_decorator.rb [hunk 13]
```diff
@@ -0,0 +1,19 @@
+class OrderAddressesDecorator < Draper::Decorator
+  delegate_all
+
+  def capitalize_address
+    "#{cast.capitalize} Address"
```
Comment (epoberezhny) [inline app/decorators/order_addresses_decorator.rb]: I18n

File: app/decorators/order_addresses_decorator.rb [hunk 14]
```diff
@@ -0,0 +1,19 @@
...
+  def full_name
+    "#{first_name} #{last_name}"
+  end
+
+  def city_zip
+    "#{city} #{zip}"
+  end
+
+  def phone_address
+    "Phone #{phone}"
```
Comment (epoberezhny) [inline app/decorators/order_addresses_decorator.rb]: I18n

File: app/decorators/order_decorator.rb [hunk 15]
```diff
@@ -0,0 +1,38 @@
+class OrderDecorator < Draper::Decorator
+  delegate_all
+  delegate :capitalize, to: :status, prefix: true
+
+  def date
+    created_at.strftime('%d/%m/%y')
```
Comment (epoberezhny) [inline app/decorators/order_decorator.rb]: I18n.l

File: app/decorators/order_decorator.rb [hunk 16]
```diff
@@ -0,0 +1,38 @@
...
+    3.times { string << '**** ' }
+    (string << credit_card.card_number.last(4)).to_s
+  end
+
+  def show_expiration_month_year
+    credit_card.expiration_month_year
+  end
+
+  def date_complete
+    completed_at.strftime('%B %-d, %Y')
```
Comment (epoberezhny) [inline app/decorators/order_decorator.rb]: I18n.l

File: spec/queries/latest_books_query_spec.rb [hunk 17]
```diff
@@ -0,0 +1,18 @@
...
+  subject(:latest_books) { described_class }
+
+  let!(:book1) { create(:book) }
+  let!(:book2) { create(:book) }
+  let!(:book3) { create(:book) }
+  let!(:book4) { create(:book) }
+
+  context 'when present latest_books' do
+    it { expect(latest_books.call.count).to eq(3) }
+    it { expect(latest_books.call).not_to be_empty }
```
Comment (epoberezhny) [inline spec/queries/latest_books_query_spec.rb]: ?

File: spec/queries/latest_books_query_spec.rb [hunk 18]
```diff
@@ -0,0 +1,18 @@
...
+    it { expect(latest_books.call).to match_array([book2, book3, book4]) }
+  end
+
+  context 'when not include in latest_books' do
+    it { expect(latest_books.call).not_to include(book1) }
+  end
```
Comment (epoberezhny) [inline spec/queries/latest_books_query_spec.rb]: ?

File: spec/queries/best_sellers_query_spec.rb [hunk 19]
```diff
@@ -0,0 +1,33 @@
...
+
+    it { expect(best_sellers.call).to match_array([book1, book4]) }
+  end
+
+  context 'when not include in best_sellers' do
+    let(:book5) { create(:book) }
+
+    it { expect(best_sellers.call).to be_empty }
+    it { expect(best_sellers.call).not_to include(book5) }
+  end
```
Comment (epoberezhny) [inline spec/queries/best_sellers_query_spec.rb]: ?

File: spec/queries/sort_books_query_spec.rb [hunk 20]
```diff
@@ -0,0 +1,62 @@
...
+  subject(:sort_books) { described_class }
+
+  let!(:book1) { create(:book, title: 'America', price: 100) }
+  let!(:book2) { create(:book, title: 'Bolivia', price: 200) }
+  let!(:book3) { create(:book, title: 'Costa Rica', price: 300) }
+  let!(:book4) { create(:book, title: 'Dominican Republic', price: 400) }
+
+  context 'when sort by default' do
+    it do
+      sort = sort_books.new(Book.all).by_sort(:newest).first
```
Comment (epoberezhny) [inline spec/queries/sort_books_query_spec.rb]: Why only first?

File: app/queries/best_sellers_query.rb [hunk 21]
```diff
@@ -1,4 +1,4 @@
-class BestSellersQuery
+class BestSellersQuery < ApplicationQuery
   DELIVERED = 3
 
   class << self
```
Comment (epoberezhny) [inline app/queries/best_sellers_query.rb]: remove this

File: app/queries/latest_books_query.rb [hunk 22]
```diff
@@ -1,4 +1,4 @@
-class LatestBooksQuery
+class LatestBooksQuery < ApplicationQuery
   ITEM_LATEST_BOOKS = 3
 
   class << self
```
Comment (epoberezhny) [inline app/queries/latest_books_query.rb]: remove this

File: app/queries/sort_books_query.rb [hunk 23]
```diff
@@ -8,13 +8,14 @@ class SortBooksQuery
     price_desc: 'price DESC'
   }.freeze
 
-  attr_accessor :initial_scope
+  attr_accessor :initial_scope, :filter
```
Comment (epoberezhny) [inline app/queries/sort_books_query.rb]: WHYYYYYYYY FILTER???????

Comment (epoberezhny) [inline app/queries/sort_books_query.rb]: WHYYYYYYYY FILTER???????

Comment (epoberezhny) [inline app/queries/sort_books_query.rb]: WHYYYYYYYY FILTER???????

File: app/services/book_service.rb [hunk 24]
```diff
@@ -6,10 +6,10 @@ def initialize(category_id, filter = nil)
     @filter = filter
   end
 
-  def filter_for_books
-    return SortBooksQuery.new(Book.all).by_sort(filter).includes(:authors_books, :authors) unless category_id
+  def sort_for_books
+    return SortBooksQuery.call(Book.all, filter).includes(:authors) unless category_id
```
Comment (epoberezhny) [inline app/services/book_service.rb]: FILTER SERIOUSLY????

File: config/application.rb [hunk 25]
```diff
@@ -7,6 +7,7 @@
 module Bookstore
   class Application < Rails::Application
     config.load_defaults 5.2
+    config.active_job.queue_adapter = :sidekiq
```
Comment (epoberezhny) [inline config/application.rb]: Should be environment specific

Comment (epoberezhny) [inline config/application.rb]: Should be environment specific

File: spec/rails_helper.rb [hunk 26]
```diff
@@ -21,11 +26,18 @@
   config.use_transactional_fixtures = false
   config.include CheckoutHelper
 
-  config.before(:suite) do
-    DatabaseCleaner.clean_with(:truncation)
+  config.before(:each) do
     DatabaseCleaner.strategy = :transaction
   end
 
+  config.before(:each, type: :feature) do
```
Comment (epoberezhny) [inline spec/rails_helper.rb]: Why it is not in support folder?

File: .ruby-version.12.18.2019-02:17:56 [hunk 27]
```diff
@@ -0,0 +1 @@
+ruby-2.5.1
```
Comment (epoberezhny) [inline .ruby-version.12.18.2019-02:17:56]: Why do we need this file?

File: app/jobs/cleaner_order_job.rb [hunk 28]
```diff
@@ -1,7 +1,6 @@
 class CleanerOrderJob < ApplicationJob
-  queue_as :default
-
   def perform(*_args)
-    Order.where(user_id: nil).destroy_all
+    Order.where(user_id: nil)
+         .where("now() >= created_at + INTERVAL '24 hours'").destroy_all
```
Comment (epoberezhny) [inline app/jobs/cleaner_order_job.rb]: maybe `where('created_at =< ?', Time.current - 1.day)` ?

File: config/routes.rb [hunk 29]
```diff
@@ -2,6 +2,10 @@
   devise_for :admin_users, ActiveAdmin::Devise.config
   ActiveAdmin.routes(self)
 
+  require 'sidekiq/web'
+  require 'sidekiq-scheduler/web'
```
Comment (epoberezhny) [inline config/routes.rb]: Move to the beginning of file

**Pull request merged**